File: sdks/typescript/test/stream-reconnect.test.ts
/**
 * What a caller draining `listen()` learns when the stream drops.
 *
 * Java's `Events` restores the subscription itself and then delivers a
 * `Reconnected`, so a consumer knows its state is stale and reseeds.
 * TypeScript delivered `StreamDropped` and stopped: the stream stayed dead
 * until someone noticed and called `fm.reconnect()`, and nothing was ever
 * delivered to say it had come back.
 *
 * `Desk` papered over this by reconnecting on `StreamDropped` itself, so
 * the gap only showed for callers using the raw callback -- which is the whole
 * point of `listen()` being public.
 *
 * Also pins the discriminants. The rename to StreamDropped/FrameUnreadable
 * changed the interface names and left `kind` reading "transport-error" and
 * "ws-exception", so the tag a caller actually switches on still spoke the old
 * vocabulary.
 */

import { test } from "node:test";
import assert from "node:assert/strict";

import { EventListener } from "../src/stomp.ts";
import type { FmEvent, Reconnected, StreamDropped } from "../src/stomp.ts";

/** An EventListener with the socket stubbed out. */
class StubListener extends EventListener {
  connects = 0;
  failConnects = 0;

  /** While held, start() parks here. Lets a test hold a reconnect open. */
  private gate: Promise<void> | null = null;
  private openGate: () => void = () => {};
  /** Resolves once a reconnect has reached the gate and is parked there. */
  reachedGate: Promise<void> = Promise.resolve();
  private markReached: () => void = () => {};

  /** Hold the next reconnect open until release() is called. */
  holdReconnect(): void {
    this.gate = new Promise<void>((resolve) => (this.openGate = resolve));
    this.reachedGate = new Promise<void>((resolve) => (this.markReached = resolve));
  }

  release(): void {
    this.openGate();
  }

  constructor(callback: (event: FmEvent) => void) {
    super(
      "ws://127.0.0.1:0/events",
      "t",
      7,
      callback,
      "stream-reconnect-test",
      (d) => d as never,
      (d) => d as never,
    );
  }

  override async start(): Promise<void> {
    this.connects += 1;
    if (this.gate) {
      this.markReached();
      await this.gate;
    }
    if (this.connects <= this.failConnects) throw new Error("refused");
  }
}

function collector() {
  const events: FmEvent[] = [];
  return { events, push: (e: FmEvent) => void events.push(e) };
}

const settle = () => new Promise((r) => setTimeout(r, 50));

/** Reject rather than hang, so a premise that never holds fails loudly. */
async function within(promise: Promise<void>, ms: number, why: string): Promise<void> {
  let timer: ReturnType<typeof setTimeout>;
  const expiry = new Promise<never>((_, reject) => {
    timer = setTimeout(() => reject(new Error(why)), ms);
  });
  try {
    await Promise.race([promise, expiry]);
  } finally {
    clearTimeout(timer!);
  }
}

test("a dropped stream restores itself and says so", async () => {
  const { events, push } = collector();
  const listener = new StubListener(push);
  await listener.start();
  const before = listener.connects;

  listener._onStreamDropped(new Error("connection reset"));
  await settle();

  assert.equal((events[0] as StreamDropped).kind, "stream-dropped");
  assert.ok(events[1], "the caller is never told the stream came back");
  assert.equal((events[1] as Reconnected).kind, "reconnected");
  assert.equal((events[1] as Reconnected).marketplaceId, 7);
  assert.ok(listener.connects > before, "nothing reconnected");
  listener.close();
});

test("a closed listener does not reconnect", async () => {
  const { events, push } = collector();
  const listener = new StubListener(push);
  await listener.start();
  listener.close();
  const before = listener.connects;

  listener._onStreamDropped(new Error("socket closed"));
  await settle();

  assert.equal(listener.connects, before);
  assert.deepEqual(events, []);
});

test("one drop starts one reconnect", async () => {
  // A burst of errors from a dying socket is one outage, not five.
  //
  // Same shape as Java's oneDropStartsOneReconnect and Python's
  // test_one_drop_starts_one_reconnect: hold the reconnect open, deliver the
  // burst, prove the window really was open, then let it finish and count.
  //
  // This used to hold the window with `failConnects = connects + 1` and a
  // 2.5s sleep, which was correct but timing-based -- and the Python copy of
  // the same idea was written `fail_connects = 1`, forgot the connect start()
  // had already spent, and silently stopped holding anything. A gate cannot
  // drift that way, and costs no wall-clock.
  const { events, push } = collector();
  const listener = new StubListener(push);
  await listener.start();
  listener.holdReconnect();

  for (let i = 0; i < 5; i++) listener._onStreamDropped(new Error("connection reset"));

  // The premise, asserted rather than assumed: a reconnect must be parked in
  // the gate right now. Remove the gate and this never resolves, so the test
  // fails every run instead of passing for the wrong reason.
  await within(
    listener.reachedGate,
    2000,
    "no reconnect reached the gate: the burst was delivered after the reconnect " +
      "had already finished, so nothing here tests the guard",
  );
  const isReconnected = (e: FmEvent) => (e as Reconnected).kind === "reconnected";
  assert.equal(events.filter(isReconnected).length, 0, "a reconnect finished before the burst was delivered");

  listener.release();
  await settle();

  const reconnected = events.filter(isReconnected);
  assert.equal(reconnected.length, 1, `expected one Reconnected, got ${reconnected.length}`);
  listener.close();
});

File: sdks/java/fm-sdk/src/test/java/fm/internal/StreamReconnectTest.java
package fm.internal;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

import fm.error.ApiException;
import fm.event.Reconnected;

/**
 * What a caller draining {@code listen()} learns when the stream drops.
 *
 * <p>Python and TypeScript have covered this all along and Java did not, which
 * is backwards: Java is where the auto-reconnect was written first, and the
 * other two were brought up to it. The behaviour is that {@link Events}
 * restores the subscription itself and then puts a {@link Reconnected} on the
 * queue, so a consumer knows its state is stale and reseeds.
 *
 * <p>It matters beyond the raw queue. {@code DefaultDesk} reseeds its books
 * from that Reconnected -- a reconnect is the largest possible sequence gap --
 * so a drop that never announced itself would leave a desk serving a stale
 * book with nothing to say so.
 *
 * <p>The socket is stubbed by overriding {@link Events#connect()}: everything
 * below the STOMP frames is what a fake would have to fake, and none of it is
 * what these are about. They are about what reaches the queue after the
 * receive loop dies.
 */
class StreamReconnectTest {

    private static final long MP = 7L;

    /** An Events with the socket stubbed out. */
    private static final class _Listener extends Events {
        int connects;
        int failConnects;

        _Listener(BlockingQueue<Object> queue) {
            super("ws://127.0.0.1:0/events", "t", MP, "fm-sdk-test",
                  HttpFlexemarkets.MAPPER, queue);
        }

        @Override
        void connect() {
            connects++;
            if (connects <= failConnects) {
                throw new ApiException("refused");
            }
        }
    }

    private static Object _poll(BlockingQueue<Object> queue) throws InterruptedException {
        return queue.poll(5, TimeUnit.SECONDS);
    }

    @Test
    @Timeout(20)
    void aDroppedStreamRestoresItselfAndSaysSo() throws Exception {
        BlockingQueue<Object> queue = new ArrayBlockingQueue<>(16);
        var listener = new _Listener(queue);

        listener.reconnectInBackground();

        assertThat(_poll(queue))
            .as("a consumer is told the stream came back, so it can reseed")
            .isInstanceOf(Reconnected.class);
        assertThat(listener.connects).isEqualTo(1);
    }

    @Test
    @Timeout(20)
    void reconnectedCarriesTheMarketplaceItBelongsTo() throws Exception {
        BlockingQueue<Object> queue = new ArrayBlockingQueue<>(16);
        var listener = new _Listener(queue);

        listener.reconnectInBackground();

        // One Flexemarkets can hold desks on several marketplaces, so a
        // Reconnected that did not name one would tell every desk to reseed.
        assertThat(_poll(queue)).isEqualTo(new Reconnected(MP));
    }

    @Test
    @Timeout(20)
    void aClosedListenerDoesNotReconnect() throws Exception {
        BlockingQueue<Object> queue = new ArrayBlockingQueue<>(16);
        var listener = new _Listener(queue);
        listener.close();

        listener.reconnectInBackground();

        assertThat(queue.poll(250, TimeUnit.MILLISECONDS))
            .as("close() is final; a late drop must not resurrect the stream")
            .isNull();
        assertThat(listener.connects).isZero();
    }

    @Test
    @Timeout(20)
    void oneDropStartsOneReconnect() throws Exception {
        BlockingQueue<Object> queue = new ArrayBlockingQueue<>(16);
        var listener = new _Listener(queue);
        // Two attempts fail, so the retry loop is still running when the
        // second and third drops arrive -- which is what a dying socket
        // actually does: onClose and onError both fire for one outage.
        listener.failConnects = 2;

        listener.reconnectInBackground();
        listener.reconnectInBackground();
        listener.reconnectInBackground();

        assertThat(_poll(queue)).isInstanceOf(Reconnected.class);
        assertThat(queue.poll(250, TimeUnit.MILLISECONDS))
            .as("a burst of errors from one dying socket is one outage, not three")
            .isNull();
    }
}
